### PR TITLE
`ManagedSeed` defaults Ingress kind to `nginx` when `seedSpec` is not specified

### LINKED ISSUE #10396: Provide proper defaulting for `ManagedSeed`,s `gardenlet.config.seedConfig.spec.ingress.controller.kind`
**How to categorize this issue?**

/area quality
/kind enhancement

**What would you like to be added**:
According to the documentation for [managed seeds](https://github.com/gardener/gardener/blob/master/docs/operations/managed_seed.md):
"You can omit practically everything from the gardenlet section, including all or most of the Seed spec fields. Proper defaults will be supplied in all cases, based either on the most common use cases or the information already available in the Shoot resource."
And when using the [example resource](https://github.com/gardener/gardener/blob/master/docs/operations/managed_seed.md#deploying-gardenlet-to-the-shoot) in a local kind cluster:
```
$ echo "
apiVersion: seedmanagement.gardener.cloud/v1alpha1
kind: ManagedSeed
metadata:
  name: my-managed-seed
  namespace: garden
spec:
  shoot:
    name: local
  gardenlet: {}
" | kubectl apply -f -
The ManagedSeed "my-managed-seed" is invalid: spec.gardenlet.config.seedConfig.spec.ingress.controller.kind: Unsupported value: "": supported values: "nginx"
```
And the following works:
```
apiVersion: seedmanagement.gardener.cloud/v1alpha1
kind: ManagedSeed
metadata:
  name: my-managed-seed
  namespace: garden
spec:
  shoot:
    name: local
  gardenlet:
    config:
      apiVersion: gardenlet.config.gardener.cloud/v1alpha1
      kind: GardenletConfiguration
      seedConfig:
        spec:
          ingress:
            controller:
              kind: nginx
```

We can validate that the change works by checking if `k -n garden get managedseeds my-managed-seed -o yaml | yq .spec` return the same when run with and without specifying `gardenlet` configuration

**Note:** After https://github.com/gardener/gardener/issues/10349 we can even drop `gardenlet: {}` from the example

**Why is this needed**:
Consistency with documentation and easier to create managed seeds with default configurations

### PR DESCRIPTION
**Pull request by RadaBDimitrova**
**How to categorize this PR?**

/area quality
/kind enhancement

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/gardener/issues/10396

**Special notes for your reviewer**:

**Release note**:

```noteworthy operator
ManagedSeed's `.spec.gardenlet.config.seedConfig.spec.ingress.controller.kind` field is now defaulted to `nginx` when
`.spec.gardenlet.config.seedConfig` or `.spec.gardenlet.config.seedConfig.spec.ingress` is nil.
 This allows the creation of ManagedSeed without specifying the `.spec.gardenlet` field.
```

### CODE REVIEW

File: pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go [hunk 1]
```diff
@@ -118,7 +118,17 @@ func setDefaultsGardenletConfiguration(obj *gardenletv1alpha1.GardenletConfigura
...
+			SeedTemplate: gardencorev1beta1.SeedTemplate{
+				Spec: gardencorev1beta1.SeedSpec{
+					Ingress: &gardencorev1beta1.Ingress{
+						Controller: gardencorev1beta1.IngressController{
+							Kind: "nginx",
+						},
+					},
+				},
+			},
+		}
```
Comment (shafeeqes) [inline pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go]: In this case, the `Ingress` field will not be defaulted if `SeedConfig` is non-nil. We should rather do
```
             if obj.SeedConfig == nil {
		obj.SeedConfig = &gardenletv1alpha1.SeedConfig{
			SeedTemplate: gardencorev1beta1.SeedTemplate{
				Spec: gardencorev1beta1.SeedSpec{},
                          },
                  }
             }
             
             if obj.SeedConfig.SeedTemplate.Spec.Ingress == nil{
                 obj.SeedConfig.SeedTemplate.Spec.Ingress = &gardencorev1beta1.Ingress{
		     Controller: gardencorev1beta1.IngressController{
			Kind: "nginx",
		      },
		  },
	      }

```

**Comment by Kostov6** [comment]
```
STEP: Verify that gardenlet is deployed @ 10/22/24 07:39:07.027
  STEP: Update some value @ 10/22/24 07:39:07.049
  STEP: Verify that value change was not rolled out @ 10/22/24 07:39:07.058
  [FAILED] in [It] - /home/prow/go/src/github.com/gardener/gardener/test/integration/operator/gardenlet/gardenlet_test.go:197 @ 10/22/24 07:39:12.058
  STEP: Delete Gardenlet @ 10/22/24 07:39:12.058
  STEP: Wait for Gardenlet to be gone @ 10/22/24 07:39:12.063
  STEP: Delete and wait for Gardenlet deployment to be gone @ 10/22/24 07:39:12.063
  << Timeline
   [FAILED] Timed out after 5.001s.
  Expected
      <int32>: 1337
  to equal
      <int32>: 3
  In [It] at: /home/prow/go/src/github.com/gardener/gardener/test/integration/operator/gardenlet/gardenlet_test.go:197 @ 10/22/24 07:39:12.058
```

File: pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go [hunk 2]
```diff
@@ -115,8 +115,21 @@ func setDefaultsGardenletConfiguration(obj *gardenletv1alpha1.GardenletConfigura
 	setDefaultsResources(obj.Resources)
 
 	// Initialize seed config
+
```
Comment (Kostov6) [inline pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go]: unneeded line

File: pkg/apis/seedmanagement/v1alpha1/defaults_gardenlet_test.go [hunk 3]
```diff
@@ -148,6 +163,11 @@ var _ = Describe("Defaults", func() {
 											Namespace: "bar",
 										},
 									},
+									Ingress: &gardencorev1beta1.Ingress{
+										Controller: gardencorev1beta1.IngressController{
+											Kind: "nginx",
+										},
+									},
```
Comment (Kostov6) [inline pkg/apis/seedmanagement/v1alpha1/defaults_gardenlet_test.go]: Lets verify that an already specified ingress configuration would not be overwritten instead

File: pkg/apis/seedmanagement/v1alpha1/defaults_managedseed_test.go [hunk 4]
```diff
@@ -166,6 +181,11 @@ var _ = Describe("Defaults", func() {
 											Namespace: "bar",
 										},
 									},
+									Ingress: &gardencorev1beta1.Ingress{
+										Controller: gardencorev1beta1.IngressController{
+											Kind: "nginx",
+										},
+									},
```
Comment (Kostov6) [inline pkg/apis/seedmanagement/v1alpha1/defaults_managedseed_test.go]: Lets verify that an already specified ingress configuration would not be overwritten instead

**Review by Kostov6 — COMMENTED** [review]
Can you check if this PR fixes https://github.com/gardener/gardener/issues/10679?

**Comment by shafeeqes** [comment]
Thanks!
/approve
/retest

**Comment by ialidzhikov** [comment]
Let me have a look into this PR. I am sorry for the delay.
/hold

File: docs/operations/managed_seed.md [hunk 5]
```diff
@@ -35,7 +35,6 @@ metadata:
 spec:
   shoot:
     name: crazy-botany
-  gardenlet: {}
```
Comment (ialidzhikov) [inline docs/operations/managed_seed.md:38]: We can also remove the following part from the example ManagedSeed:
```diff
diff --git a/example/provider-local/managedseeds/managedseed.yaml b/example/provider-local/managedseeds/managedseed.yaml
index 20f3949a2..888a42c66 100644
--- a/example/provider-local/managedseeds/managedseed.yaml
+++ b/example/provider-local/managedseeds/managedseed.yaml
@@ -19,6 +19,3 @@ spec:
               visible: false
             verticalPodAutoscaler:
               enabled: false
-          ingress:
-            controller:
-              kind: nginx
```

File: pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go [hunk 6]
```diff
@@ -116,7 +116,19 @@ func setDefaultsGardenletConfiguration(obj *gardenletv1alpha1.GardenletConfigura
 
 	// Initialize seed config
 	if obj.SeedConfig == nil {
-		obj.SeedConfig = &gardenletv1alpha1.SeedConfig{}
+		obj.SeedConfig = &gardenletv1alpha1.SeedConfig{
+			SeedTemplate: gardencorev1beta1.SeedTemplate{
+				Spec: gardencorev1beta1.SeedSpec{},
+			},
+		}
```
Comment (ialidzhikov) [inline pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go]: I guess we can revert this change. 

```golang
		obj.SeedConfig = &gardenletv1alpha1.SeedConfig{
			SeedTemplate: gardencorev1beta1.SeedTemplate{
				Spec: gardencorev1beta1.SeedSpec{},
			},
		}
```

should be equivalent to 

```golang
		obj.SeedConfig = &gardenletv1alpha1.SeedConfig{}
```

The `SeedTemplate` and its `Spec` fields are not nil-able. So, we don't have to initialize them to avoid nil pointer.

Or is there any other reason for this change that I am missing?

File: pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go [hunk 7]
```diff
@@ -116,7 +116,19 @@ func setDefaultsGardenletConfiguration(obj *gardenletv1alpha1.GardenletConfigura
...
+	}
+
+	if obj.SeedConfig.SeedTemplate.Spec.Ingress == nil {
+		obj.SeedConfig.SeedTemplate.Spec.Ingress = &gardencorev1beta1.Ingress{
+			Controller: gardencorev1beta1.IngressController{
+				Kind: "nginx",
```
Comment (ialidzhikov) [inline pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go]:
```suggestion
				Kind: v1beta1constants.IngressKindNginx,
```

Comment (shafeeqes) [inline pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go]: Can we use the pre defined constant here?
https://github.com/gardener/gardener/blob/d3eacb5e15756a457c5fc232a3963fb62424a86f/pkg/apis/core/v1beta1/constants/types_constants.go#L750-L751
Also, the doc string of this constant doesn't sound correct. Could you please correct that as well?

File: pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go [hunk 8]
```diff
@@ -116,7 +116,19 @@ func setDefaultsGardenletConfiguration(obj *gardenletv1alpha1.GardenletConfigura
...
+			},
+		}
```
Comment (ialidzhikov) [inline pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go]: I am not sure if this defaulting belongs here. The `ingress.domain` field is being defaulted/initialized in the validator admission plugin: https://github.com/gardener/gardener/blob/024b9f343303febe4a8def668dc55f6eb78b812b/plugin/pkg/managedseed/validator/admission.go#L371-L376

Defaulting here only the ingress controller kind and the domain to be defaulted somewhere else does not make sense. I guess the ingress kind also have to be defaulted in the admission plugin? WDYT?

Comment (ialidzhikov) [inline pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go:131]: I would expect to see a test that verifies that you are now able to create a ManagedSeed without specifying `.spec.gardenlet` (the real motivation for the issue this PR is fixing). If we address the previous comment, then we can add it as an unit test to the admission plugin.

Comment (ialidzhikov) [reply pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go:131]: OK, when thinking about it again let's keep it as it is.

Comment (Kostov6) [reply pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go]: In terms of code locality your suggestion makes sense. But conceptually what is the purpose of validator admission plugin? Why is there defaulting logic in the first place? What is the context in which you want to disable it and the defaulting in it not to happen?

Comment (ialidzhikov) [reply pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go]: > But conceptually what is the purpose of validator admission plugin?

In admission plugins we usually do validations (the cross-object validation, not the static one) and defaulting. The defaulting is cross-object in our particular case - we have to fetch the Shoot resource to compute the ManagedSeed's ingress domain and later on to validate it. Historically, we don't do this operation in the static validation or defaulting for the ManagedSeed as there referenced resource is not available there.

> Why is there defaulting logic in the first place?

I am also not very happing that admission plugins with validator in their name are doing defaulting as well. Years ago I created https://github.com/gardener/gardener/issues/2158 but I guess no one had capacity to work on it.

> What is the context in which you want to disable it and the defaulting in it not to happen?

I don't have any productive use case in might. You see that in tests this is the case. And theoretically you are allowed to do it. 

---

Maybe indeed I would move the defaulting to the admission plugin because:
- the current logic for setting the ingress field in the admission plugin is incomplete. It requires right now something or someone to specify the controller.kind field
- the current defaulting in this PR is incomplete as it requires something or someone to specify the ingress.domain field

WDYT?

Comment (Kostov6) [reply pkg/apis/seedmanagement/v1alpha1/defaults_managedseed.go]: Your suggestion will increase code locality and there are no apparent drawbacks so I agree with it

**Comment by RadaBDimitrova** [comment]
Known flake #10576
/retest

File: test/integration/gardenlet/managedseed/managedseed_test.go [hunk 9]
```diff
@@ -112,6 +112,15 @@ var _ = Describe("ManagedSeed controller test", func() {
 						},
 					},
 					Spec: gardencorev1beta1.SeedSpec{
+						DNS: gardencorev1beta1.SeedDNS{
```
Comment (ialidzhikov) [inline test/integration/gardenlet/managedseed/managedseed_test.go]: As the defaulting is now moved to the admission plugin and the admission plugin is disabled for this integration test, you should be able to revert the changes from this file.

File: test/integration/gardenlet/gardenlet/gardenlet_test.go [hunk 10]
```diff
@@ -47,10 +47,25 @@ var _ = Describe("Gardenlet controller test", func() {
 						},
 					},
 					Spec: gardencorev1beta1.SeedSpec{
+						DNS: gardencorev1beta1.SeedDNS{
```
Comment (ialidzhikov) [inline test/integration/gardenlet/gardenlet/gardenlet_test.go]: As the defaulting is now moved to the admission plugin and the admission plugin is disabled for this integration test, you should be able to revert the changes from this file.

File: test/integration/controllermanager/shoot/conditions/conditions_test.go [hunk 11]
```diff
@@ -70,6 +72,33 @@ var _ = Describe("Shoot Conditions controller tests", func() {
 			Expect(client.IgnoreNotFound(testClient.Delete(ctx, shoot))).To(Succeed())
 		})
 
+		gardenletConfig := &gardenletv1alpha1.GardenletConfiguration{
```
Comment (ialidzhikov) [inline test/integration/controllermanager/shoot/conditions/conditions_test.go]: As the defaulting is now moved to the admission plugin and the admission plugin is disabled for this integration test, you should be able to revert the changes from this file.

File: test/integration/controllermanager/seed/lifecycle/lifecycle_test.go [hunk 12]
```diff
@@ -102,15 +104,44 @@ var _ = Describe("Seed Lifecycle controller tests", func() {
 			Expect(testClient.Delete(ctx, seed)).To(Succeed())
 		})
 
+		gardenletConfig := &gardenletv1alpha1.GardenletConfiguration{
```
Comment (ialidzhikov) [inline test/integration/controllermanager/seed/lifecycle/lifecycle_test.go]: As the defaulting is now moved to the admission plugin and the admission plugin is disabled for this integration test, you should be able to revert the changes from this file.

File: plugin/pkg/managedseed/validator/admission_test.go [hunk 13]
```diff
@@ -374,7 +374,7 @@ var _ = Describe("ManagedSeed", func() {
 
 				err := admissionHandler.Admit(context.TODO(), getManagedSeedAttributes(managedSeed), nil)
 				Expect(err).NotTo(HaveOccurred())
-
+				seedx.Spec.Ingress.Controller.Kind = v1beta1constants.IngressKindNginx
```
Comment (ialidzhikov) [inline plugin/pkg/managedseed/validator/admission_test.go:378]: Cosmetic nit: We could keep the empty line to visually separate the different phases - setup, invocation of the method under test, assertion.

```suggestion

				seedx.Spec.Ingress.Controller.Kind = v1beta1constants.IngressKindNginx
```

**Comment by RadaBDimitrova** [comment]
Known flake #10576
/retest

**Review by ialidzhikov — COMMENTED** [review]
/approve

**Review by Kostov6 — COMMENTED** [review]
/lgtm
